Refactor actor reminders and timers into their own packages

DeepanshuA opened the pull request:
# Description

Segregates Reminders and Timers interface from Actors package.

## Issue reference

<!--
We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.
-->

Please reference the issue this PR will close: #6591 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

ItalyPaleAle commented:
/ok-to-test

ItalyPaleAle commented on pkg/actors/reminders/reminders.go:
```diff
@@ -0,0 +1,975 @@
+	"time"
+
+	"k8s.io/utils/clock"
+
```
Nit: extra newline here :)

DeepanshuA replied:
Done

ItalyPaleAle commented on pkg/actors/reminders/reminders.go:
```diff
@@ -0,0 +1,975 @@
+)
+
+// Implements a reminders provider.
+//
```
Nit: please remove these 2 empty lines

DeepanshuA replied:
Done

ItalyPaleAle commented on pkg/actors/reminders/reminders.go:
```diff
@@ -0,0 +1,975 @@
+	r.remindersLock.RUnlock()
+
+	for _, rem := range reminders {
+		// r.reminder refers to the actual reminder struct that is saved in the db
```
What's r.reminder here?

DeepanshuA replied:
Yup, replaced it with `rem.Reminder`. That should be the new one.

ItalyPaleAle commented on pkg/actors/reminders/reminders.go:
```diff
@@ -0,0 +1,975 @@
+}
+
+func (r *reminders) RenameReminder(ctx context.Context, req *internal.RenameReminderRequest) error {
+	log.Warn("[DEPRECATION NOTICE] Currently RenameReminder renames by deleting-then-inserting-again. This implementation is not fault-tolerant, as a failed insert after deletion would result in no reminder")
```
Looks like this was deprecated almost 1 year ago in Dapr 1.9 (#4957). Is it time we remove this method maybe?
CC: @dapr/maintainers-dapr

yaron2 replied:
Yes, lets remove this

ItalyPaleAle replied:
I've opened #6711 to track this.

ItalyPaleAle commented on pkg/actors/reminders/reminders_test.go:
```diff
@@ -0,0 +1,1747 @@
+}
+
+func fakeTStore() (internal.TransactionalStateStore, error) {
+	return nil, errors.New(errStateStoreNotConfigured)
```
I am not sure I understand the purpose of this. Providing a state store that always fails? Also, why `errStateStoreNotConfigured`?

DeepanshuA replied:
I changed the name to `fakeNoStore`. Basically, this is required in `TestStoreIsNotInitialized`, when state store is not fetched properly.

ItalyPaleAle commented on pkg/actors/reminders/reminders_test.go:
```diff
@@ -0,0 +1,1747 @@
+	return nil, errors.New(errStateStoreNotConfigured)
+}
+
+var fakeStor internal.TransactionalStateStore
+
+func fakeRealTStore() (internal.TransactionalStateStore, error) {
+	if fakeStor == nil {
+		fakeStor = daprt.NewFakeStateStore()
+	}
+	return fakeStor, nil
+}
+
+func fakeRealTStore2() (internal.TransactionalStateStore, error) {
+	return daprt.NewFakeStateStore(), nil
+}
```
Why these two "fake real" (quite the oxymoron!) stores?

Also, why is `fakeStor` initialized lazily? That may cause race conditions

DeepanshuA replied:
Removed one of them - was not required, seems left while testing something.
Removed lazy loading of `fakeStore` too.

ItalyPaleAle commented on pkg/actors/reminders/reminders_test.go:
```diff
@@ -0,0 +1,1747 @@
+	testReminders.SetLookupActorFn(func(string, string) (bool, string) {
+		return true, localhost
+	})
+	testReminders.Init(context.TODO())
```
In test code, we use `context.Background()` becaus it's not a TODO to fix in the future.

DeepanshuA replied:
Good point - corrected all.

ItalyPaleAle commented on pkg/actors/actors.go:
```diff
@@ -98,73 +98,41 @@ type PlacementService interface {
+	timers               internal.TimersProvider
+	actorsReminders      internal.RemindersProvider
+	actorsTable          *sync.Map
+	evaluationChan       chan struct{}
```
Looks like this is unused now

DeepanshuA replied:
Yeah, removed.

ItalyPaleAle commented on pkg/actors/actors.go:
```diff
@@ -187,20 +155,25 @@ func newActorsWithClock(opts ActorsOpts, clock clock.WithTicker) Actors {
 		grpcConnectionFn:     opts.GRPCConnectionFn,
-		config:               opts.Config,
+		config:               opts.ActorsConfig,
+		coreConfig:           &opts.ActorsConfig.Config,
```
If "coreConfig" is part of `opts.ActorConfig`, why do we need to store them separately in the object? Can't we just store `config: opts.ActorsConfig` and then access the `Config` property of that?

DeepanshuA replied:
Had placed it, as had to refer everywhere `config.Config`.
Removed it now as per the comment, and using `actorsConfig.Config`

ItalyPaleAle commented on pkg/actors/actors.go:
```diff
@@ -765,243 +748,43 @@ func (a *actorsRuntime) drainRebalancedActors() {
 
 // Executes a reminder or timer
-func (a *actorsRuntime) executeReminder(reminder *reminders.Reminder, isTimer bool) (err error) {
+func (a *actorsRuntime) executeReminder(reminder *internal.Reminder, isTimer bool) (err error) {
```
Nit on naming: perhaps this can be called `doExecuteReminderOrTimer` (or something like that), so we can have `executeTimer` and `executeReminder` above (instead of `executeReminderWrapper`). Or maybe both "wrapper" functions could be called like `executeTimerFn` and `executeReminderFn`? Anything as long as it's consistent.

DeepanshuA replied:
Corrected.

ItalyPaleAle reviewed (changes requested):
Thanks @DeepanshuA for this PR. It's not easy to "mess" with the actors subsystem, but this looks very good. My comments on the code are fairly small, especially considering what this code is about.

It's great that E2E tests are passing consistently.

However, it seems that the unit tests are a bit flaky right now, for example: https://github.com/dapr/dapr/actions/runs/5600637276/jobs/10243511144?pr=6669
This may be a symptom of race conditions, and running the tests with `go test -race` may help.

DeepanshuA commented:
> Thanks @DeepanshuA for this PR. It's not easy to "mess" with the actors subsystem, but this looks very good. My comments on the code are fairly small, especially considering what this code is about.
> 
> It's great that E2E tests are passing consistently.
> 
> However, it seems that the unit tests are a bit flaky right now, for example: https://github.com/dapr/dapr/actions/runs/5600637276/jobs/10243511144?pr=6669 This may be a symptom of race conditions, and running the tests with `go test -race` may help.

Thanks for the quick review @ItalyPaleAle. Have addressed review comments, will get to the flakiness as well.

JoshVanL commented on pkg/actors/actors.go:
```diff
@@ -86,86 +91,38 @@ type Actors interface {
 type ActorsOpts struct {
 	AppChannel       channel.AppChannel
 	GRPCConnectionFn GRPCConnectionFn
-	Config           Config
+	ActorsConfig     Config
```
I think keeping this `Config` rather than `ActorsConfig` since `Actors` is implied by the package and  type.

ItalyPaleAle replied:
Done

JoshVanL reviewed (changes requested):
Couple nits from me, but otherwise LGTM! :slightly_smiling_face:

ItalyPaleAle commented on pkg/actors/reminders/actor_metadata.go:
```diff
@@ -0,0 +1,159 @@
+/*
```
Done
